apps/cms/src/pages/api/webhooks/product-updated.ts

```diff
@@ -0,0 +1,127 @@
+import { NextWebhookApiHandler, SaleorAsyncWebhook } from "@saleor/app-sdk/handlers/next";
+import { gql } from "urql";
+import {
+  ProductUpdatedWebhookPayloadFragment,
+  UntypedWebhookProductFragmentDoc,
+} from "../../../../generated/graphql";
+import { saleorApp } from "../../../../saleor-app";
+import { getCmsKeysFromSaleorItem } from "../../../lib/cms/client/metadata";
+import { getChannelsSlugsFromSaleorItem } from "../../../lib/cms/client/channels";
+import {
+  createCmsOperations,
+  executeCmsOperations,
+  executeMetadataUpdate,
+} from "../../../lib/cms/client";
+import { logger as pinoLogger } from "../../../lib/logger";
+import { createClient } from "../../../lib/graphql";
+import { fetchProductVariantMetadata } from "../../../lib/metadata";
+
+export const config = {
+  api: {
+    bodyParser: false,
+  },
+};
+
+export const ProductUpdatedWebhookPayload = gql`
+  ${UntypedWebhookProductFragmentDoc}
+  fragment ProductUpdatedWebhookPayload on ProductUpdated {
+    product {
+      ...WebhookProduct
+    }
+  }
+`;
+
+export const ProductUpdatedSubscription = gql`
+  ${ProductUpdatedWebhookPayload}
+  subscription ProductUpdated {
+    event {
+      ...ProductUpdatedWebhookPayload
+    }
+  }
+`;
+
+export const productUpdatedWebhook = new SaleorAsyncWebhook<ProductUpdatedWebhookPayloadFragment>({
+  name: "Cms-hub product updated webhook",
+  webhookPath: "api/webhooks/product-updated",
+  event: "PRODUCT_UPDATED",
+  apl: saleorApp.apl,
+  query: ProductUpdatedSubscription,
+});
+
+export const handler: NextWebhookApiHandler<ProductUpdatedWebhookPayloadFragment> = async (
+  req,
+  res,
+  context
+) => {
+  const { product } = context.payload;
+  const { saleorApiUrl, token } = context.authData;
+
+  const logger = pinoLogger.child({
+    product,
+  });
+  logger.debug("Called webhook PRODUCT_UPDATED");
+
+  if (!product) {
+    return res.status(500).json({
+      errors: [
+        "No product product data payload provided. Cannot process product variants syncronisation in CMS providers.",
+      ],
+    });
+  }
+
+  const client = createClient(saleorApiUrl, async () => ({
+    token: token,
+  }));
+
+  const allCMSErrors: string[] = [];
+
+  product.variants?.forEach(async (variant) => {
+    const { variants: _, ...productFields } = product;
+    const productVariant = {
+      product: productFields,
+      ...variant,
+    };
+
+    const productVariantChannels = getChannelsSlugsFromSaleorItem(productVariant);
+    const productMetadata = await fetchProductVariantMetadata(client, productVariant.id);
```

Comment:
Can't it be queried with the subscription?

Reply:
Unfortunately no, because metadata might became stale for instance in the case when webhook event arrived with CMS ids not yet saved in metadata, while being in the meantime modified by other webhook handler, that saved CMS id in metadata. It caused double creation of product variant in CMS providers, because here handler do not know anything about saved CMS ids in the meantime. Therefore I added fetching latest metadata (that have CMS ids) once webhook event is handled.

Reply:
It works with current Saleor Dashboard, but it's not ideal, as potentially webhooks might be triggered parallelly (am I right?), so still potential race condition might occur. I suppose it could be solved by implementing queues or somewhat different?

Reply:
Could be solved by marking products as changed and checking which should be updated periodically on the worker instance. It would debounce multiple duplicated requests and reduce the issue.

Lets leave the current implementation then!